convert2rhel/pkghandler.py

```diff
@@ -1075,3 +1082,45 @@ def _get_packages_to_update_dnf():
         packages.append(package.name)
 
     return packages
+
+
+def package_version_cmp(pkg_1, pkg_2):
```

Comment:
Hmmm... A  few higher level thoughts on this function:

* Sometimes yum and dnf return NEVR like this: `NetworkManager-1:1.18.8-2.0.1.el7_9.x86_64`.  Other times they return it like this (ENVR form): `1:NetworkManager-1.18.8-2.0.1.el7_9.x86_64`.  Should we handle both of these cases?  In one function or in two?

* There's already a `pkghandler.compare_package_versions` function.  That one takes its input as EVR strings: `1:1.18.8-2.0.1.el7_9.x86_64` which removes the ambiguity around whether the epoch is before or after name because there is no name.  However, we can't use that directly as a key function (since the key will be a NEVR or an ENVR string).

* Should we verify that the name is the same for both strings passed into this function?  If I call the function with `package_version_cmp("test-1.0-1", "foobar-2.0-1") I would expect it to error rather than say foobar is later than test.

* Should we parse arch off the end of the NEVR (arch is `x86_64`, `i686`, `s390x`, etc) and error if arches do not match?

Here's my strawman to address all of those:

* Separate version parsing and version comparison
* The parsing function will take a version string in either ENVR or NEVR format. 
  * If epoch or arch does not exist in the input string, it will be set to `None` in the tuple.
  * Parsing arch will be slightly tricky as I don't think there's a fool-proof way to separate an optional arch from release.  Maybe looking for a list of allowed arch patterns is the best we can do. Something like: `.*(x86_64|s390x|i.86)?$`
* The version comparison function will take two version strings in either ENVR or NEVR format.
* The strings will be passed to the parsing functions to return the information in a normalized form.
  * The names and arches will be compared. 
    * An error will be emitted if names do not match.
    * An error will be emitted if arches do not match and both arches are not None.
* rpm.labelCompare() will then be used to compare the EVR for both packages and the results returned.
* Port existing use of `pkghandler.compare_package_versions()` to use the new comparison function.
* Remove `pkghandler.compare_package_versions()` and `utils.string_to_version()`.

Discuss?

Reply:
resolving this one since it's a duplicate.